Send more metrics from broadcasting node

darkdarkdragon opened the pull request:
@j0sh @ericxtang 
please review, I think it's almost complete, only left to add code that changes endpoint based on network

darkdarkdragon committed:
Send more metrics from broadcasting node

darkdarkdragon commented:
@j0sh I've added code that chooses different endpoints based on network, so I think it's good to go

darkdarkdragon committed:
Send nonce along with LogStartBroadcastClientFailed event

j0sh commented on server/rpc.go:
```diff
@@ -437,27 +438,36 @@ func StartBroadcastClient(orchestratorServer string, node *core.LivepeerNode, jo
 }
 
-func SubmitSegment(bcast Broadcaster, seg *stream.HLSSegment) {
+func SubmitSegment(bcast Broadcaster, seg *stream.HLSSegment, nonce uint64) {
```
For future improvement, the nonce could be put into the Broadcaster since it doesn't change on a segment-by-segment basis

j0sh commented on monitor/events.go:
```diff
@@ -3,76 +3,263 @@ package monitor
+
+func SetNetwork(n string) {
+	network = n
+	eventsURL = eventsURLBase + n + "/events"
```
I'm not crazy about the `SetNetwork` approach because it forces a specific URL structure on whoever may want to run a metrics system, which this means it *needs* to be either `rinkeby/events` or `mainnet/events` and so forth. What about dev env or custom environments?

A cleaner approach seems to be to set the complete URL base within livepeernode, and remove the SetNetwork function.

darkdarkdragon replied:
Yes, we should enforce url structure, either in events.go or in livepeer.go - or else someone could easily run node on rinkeby and send metrics to the /mainnet endpoint. Or else we should send `network` field with each event

j0sh replied:
Yeah that's possible, but we're also banking a lot on trust with the metrics system, since anyone can modify anything. We can hard-code "http://metrics.livepeer.org/v2/rinkeby/events" or "http://metrics.livepeer.org/v2/mainnet/events" as the monhost. See how it's done for Infura.

https://github.com/livepeer/go-livepeer/blob/2e9b5dacb961d2027d189c7603bdf07220dce121/cmd/livepeer/livepeer.go#L107-L117

darkdarkdragon replied:
My point is that when we are not enforcing url scheme we are unnecessarily creating possibility for errors in configuration - users will not to remember that two commandline arguments should be change together.
Also, same scheme is enforced on receiving side - so, if user will want to have receiving url not to be `xxx/(mainnet|rinkeby)/events` - he will need to change source code of metrics server anyway, and if he go thus far, he will change goclient's code without problem

j0sh replied:
They shouldn't have to set the URL themselves as long as `-monitor` is true. I can see `/events` being hardcoded on the metrics server, but I don't think the network name should be.

j0sh commented on monitor/events.go:
```diff
@@ -3,76 +3,263 @@ package monitor
+
+func SegmentUploadStart(nonce, seqNo uint64) {
+	if metrics.lastSegmentNonce == nonce {
+		metrics.segmentsInFlight++
```
This is OK for now, we can keep track of it for now. Eventually I'd like to see if we can just calculate the same data reliably through server-side analytics and keep the client-side metrics collection super simple, like a time series.

j0sh commented on monitor/events.go:
```diff
@@ -3,76 +3,263 @@ package monitor
+		seqDif := int64(seqNo) - metrics.lastSeqNo
+		metrics.lastSeqNo = int64(seqNo)
+		if seqDif != 1 {
+			e.Properties["seqNoDif"] = seqDif
```
Likewise, this is the sort of thing that could be done on the server to reduce local bookkeeping. But we can keep this here for now and compare results.

j0sh commented on monitor/events.go:
```diff
@@ -3,76 +3,263 @@ package monitor
+	sendPost(e)
+}
+
+func sendPost(e *event) {
```
This could be refactored into `sendPost(name string, nonce uint64, props map[string]interface{})` which would save the event-creation boilerplate in each function

j0sh reviewed (changes requested):
Looking good, getting very close. Only requesting a couple changes; one is to eliminate SetNetwork and just set the full base path through `monhost` , and to refactor sendPost to reduce some of the boilerplate around constructing each event. The rest are thoughts for the future.

darkdarkdragon committed:
Remove SetNetwork, specify full url in livepeer.go
Remove boilerplate code when creating metrics events

darkdarkdragon commented:
@j0sh I've made requested changes, will change urls after making changes to metrics node part

j0sh commented:
Looks great! 🚢

darkdarkdragon committed:
remove code duplication, not send segmentsInFlight metric if zero

darkdarkdragon commented:
@j0sh we anyway need checks for nil on server everywhere, or one malformed message will bring whole node down

darkdarkdragon committed:
check for response code after submitting segment

darkdarkdragon committed:
read body of http request in case status code is not 200

j0sh commented on server/rpc.go:
```diff
@@ -467,18 +477,52 @@ func SubmitSegment(bcast Broadcaster, seg *stream.HLSSegment) {
+		}
+		return
+	}
+	if resp.StatusCode != 200 {
```
Maybe we could do this check after the response body has been read, since we read the body in both the 200 and non-200 cases?

darkdarkdragon commented:
No, because in case of error body returned immediately, but in case all ok
- we must send “SegmentUploaded” event before starting to read body -
because body will be returned to us only after transcode

вт, 7 серп. 2018 о 22:53 Josh Allmann <notifications@github.com> пише:

> *@j0sh* commented on this pull request.
> ------------------------------
>
> In server/rpc.go
> <https://github.com/livepeer/go-livepeer/pull/501#discussion_r208363815>:
>
> >  	if err != nil {
>  		glog.Error("Unable to submit segment ", seg.SeqNo, err)
> +		if monitor.Enabled {
> +			monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, err.Error())
> +		}
> +		return
> +	}
> +	if resp.StatusCode != 200 {
>
> Maybe we could do this check after the response body has been read, since
> we read the body in both the 200 and non-200 cases?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/livepeer/go-livepeer/pull/501#pullrequestreview-144158209>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB8OnSOcHI-v5YfX3yax6TJ36lIBHK-Mks5uOfA_gaJpZM4VtBbc>
> .
>

darkdarkdragon committed:
increase grpc timeout to 8 seconds
propagate errors from transcoder to broadcaster

darkdarkdragon commented:
@j0sh please review my last commit. I've changed behavior of Orchestrator.GetJob - I think it's ok, but maybe I've missed something

j0sh commented on server/rpc.go:
```diff
@@ -180,20 +180,20 @@ func verifyMsgSig(addr ethcommon.Address, msg string, sig []byte) bool {
 }
 
-func verifyTranscoderReq(orch Orchestrator, req *TranscoderRequest, job *lpTypes.Job) bool {
+func verifyTranscoderReq(orch Orchestrator, req *TranscoderRequest, job *lpTypes.Job) (bool, error) {
```
We could probably simplify to just return error here? Also, the RPC tests probably need to be updated to match the new signature.

j0sh commented on server/rpc.go:
```diff
@@ -285,10 +285,10 @@ func GetTranscoder(context context.Context, orch Orchestrator, req *TranscoderRe
-	if !verifyTranscoderReq(orch, req, job) {
-		return nil, fmt.Errorf("Invalid transcoder request")
+	if ok, err := verifyTranscoderReq(orch, req, job); !ok || err != nil {
+		return nil, fmt.Errorf("Invalid transcoder request (%v)", err)
```
Normally I would just `return nil, err` here for brevity since we know that the error is wrapped within `verifyTranscoderReq`, but that's up to you. Concatenating the errors could be somewhat useful to build a pseudo-backtrace if we ever need to call this function in several places.

j0sh commented on server/rpc.go:
```diff
@@ -77,7 +77,7 @@ func (orch *orchestrator) GetJob(jid int64) (*lpTypes.Job, error) {
 		if err != nil {
 			glog.Errorf("Could not get assigned transcoder for job %v, err: %s", jid, err.Error())
-			// continue here without a valid transcoder address
+			return nil, err
```
I'm a bit unsure about this one; it seems like it would lead to unintended side effects.

For example, we're OK if there's a spurious error in acquiring the transcoder address during `verifyToken` , since the token has already been signed by the orchestrator (so we assume the job is valid). The job is only needed on the transcoder to ensure it hasn't expired. Returning an error here would prevent the transcoder from processing a segment, when we still want it to process the segment.

Note that we already already check if the assigned transcoder does not match the current transcoder in `verifyTranscoderReq`. If that check passes on the orchestrator, the orchestrator signs a token which the transcoder trusts.

darkdarkdragon commented:
@j0sh ok, I've revert GetJob behavior change, changed `verifyTranscoderReq` signature and fixed tests
I'm still concerned about error not getting propagated, but anyway this code is subject for change, so it is ok for now, I think

j0sh commented:
LGTM

darkdarkdragon committed:
point default endpoints for metrics to http://metrics-rinkeby.livepeer.org/ and http://metrics-mainnet.livepeer.org/
